{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b29ebd29",
   "metadata": {},
   "source": [
    "# libraries and files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "40da7de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "library(GenomicDataCommons)\n",
    "library(TCGAutils)\n",
    "library(TCGAbiolinks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d95f0719",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'sets':\n",
      "  method        from   \n",
      "  print.element ggplot2\n",
      "\n",
      "\n",
      "Attaching package: ‘sets’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:msigdbr’:\n",
      "\n",
      "    %>%\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:magrittr’:\n",
      "\n",
      "    %>%\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    set\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'enrichplot':\n",
      "  method               from\n",
      "  fortify.enrichResult DOSE\n",
      "\n",
      "clusterProfiler v3.14.3  For help: https://guangchuangyu.github.io/software/clusterProfiler\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "Guangchuang Yu, Li-Gen Wang, Yanyan Han, Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 2012, 16(5):284-287.\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sets’:\n",
      "\n",
      "    %>%\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicDataCommons’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:magrittr’:\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ggpubr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sets’:\n",
      "\n",
      "    %>%\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicDataCommons’:\n",
      "\n",
      "    facet\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# non-oncogenic gene sets\n",
    "library(msigdbr)\n",
    "library(ggplot2)\n",
    "library(stats)\n",
    "library(sets)\n",
    "library(biomaRt)\n",
    "library(clusterProfiler)\n",
    "library(data.table)\n",
    "library(stringi)\n",
    "library(pheatmap)\n",
    "library(tidyr)\n",
    "library(ggpubr)\n",
    "library(biomaRt)\n",
    "human <- readRDS(\"/data/timonaj/biomart_orthologs/human.RDS\")\n",
    "\n",
    "# go\n",
    "go_gene_sets <- msigdbr(species = \"Homo sapiens\", category = \"C5\")\n",
    "unique_go_genes <- unique(go_gene_sets$gene_symbol)\n",
    "\n",
    "# kegg\n",
    "kegg_gene_sets <- msigdbr(species = \"Homo sapiens\", category = \"C2\", subcategory = \"KEGG\")\n",
    "unique_kegg_genes <- unique(kegg_gene_sets$gene_symbol)\n",
    "\n",
    "# oncogenic\n",
    "\n",
    "## hallmark\n",
    "H_gene_sets <- msigdbr(species = \"Homo sapiens\", category = \"H\")\n",
    "unique_H_genes <- unique(H_gene_sets$gene_symbol)\n",
    "\n",
    "## C2\n",
    "c2_gene_sets <- msigdbr(species = \"Homo sapiens\", category = \"C2\")\n",
    "unique_c2_genes <- unique(c2_gene_sets$gene_symbol)\n",
    "\n",
    "\n",
    "## C4 CM\n",
    "c4_gene_sets <- msigdbr(species = \"Homo sapiens\", category = \"C4\", subcategory = \"CM\")\n",
    "unique_c4_genes <- unique(c4_gene_sets$gene_symbol)\n",
    "\n",
    "## cosmic\n",
    "cosmic_set <- fread(\"cancer_gene_census.csv\")\n",
    "unique_cosmic_genes <- unique(cosmic_set$`Gene Symbol`)\n",
    "\n",
    "# Background set of genes\n",
    "background_set <- fread(\"background_set.txt\")\n",
    "\n",
    "pathways <- readRDS(\"pathways.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21a01465",
   "metadata": {},
   "source": [
    "# data download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ac541d01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"download completed in\"\n",
      "   user  system elapsed \n",
      " 35.022  11.421  52.976 \n"
     ]
    }
   ],
   "source": [
    "### files to be downloaded\n",
    "\n",
    "# read in the .txt file for tissue type as well as the larger gene expression file\\\n",
    "ptm <- proc.time()\n",
    "gene_tpm <- fread(file = \"tcga_RSEM_gene_tpm\")\n",
    "gene_tpm_copy <- as.data.frame(gene_tpm)\n",
    "print(\"download completed in\")\n",
    "print(proc.time() - ptm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cca7160e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# length of the numerical tpm values since the first 2 columns are characters\n",
    "tpm_length <- 2:length(gene_tpm_copy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a607422a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>60498</li><li>10536</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 60498\n",
       "\\item 10536\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 60498\n",
       "2. 10536\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 60498 10536"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(gene_tpm_copy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eec4bbb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 10536</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>sample</th><th scope=col>TCGA-19-1787-01</th><th scope=col>TCGA-S9-A7J2-01</th><th scope=col>TCGA-G3-A3CH-11</th><th scope=col>TCGA-EK-A2RE-01</th><th scope=col>TCGA-44-6778-01</th><th scope=col>TCGA-F4-6854-01</th><th scope=col>TCGA-AB-2863-03</th><th scope=col>TCGA-C8-A1HL-01</th><th scope=col>TCGA-EW-A2FS-01</th><th scope=col>⋯</th><th scope=col>TCGA-DJ-A2QC-01</th><th scope=col>TCGA-A8-A09K-01</th><th scope=col>TCGA-61-1907-01</th><th scope=col>TCGA-IB-7885-01</th><th scope=col>TCGA-95-7947-01</th><th scope=col>TCGA-VQ-AA6F-01</th><th scope=col>TCGA-BR-8588-01</th><th scope=col>TCGA-24-2254-01</th><th scope=col>TCGA-DD-A115-01</th><th scope=col>TCGA-FV-A3I0-11</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000242268.2 </td><td>-9.9658</td><td> 0.2998</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>⋯</td><td>-1.3921</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-1.9379</td><td>-9.9658</td><td>-9.9658</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000259041.1 </td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>⋯</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENSG00000270112.3 </td><td>-3.8160</td><td>-3.0469</td><td>-9.9658</td><td>-9.9658</td><td>-5.5735</td><td>-9.9658</td><td>-3.4580</td><td>-9.9658</td><td>-5.0116</td><td>⋯</td><td>-6.5064</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-6.5064</td><td>-9.9658</td><td>-9.9658</td><td>-5.5735</td><td>-9.9658</td><td>-9.9658</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENSG00000167578.16</td><td> 5.2998</td><td> 4.8881</td><td> 3.5572</td><td> 4.2563</td><td> 5.3162</td><td> 4.5161</td><td> 3.6242</td><td> 4.9782</td><td> 5.7035</td><td>⋯</td><td> 5.4591</td><td> 5.5364</td><td> 4.7798</td><td> 5.0514</td><td> 6.1607</td><td> 5.2814</td><td> 3.9599</td><td> 5.0760</td><td> 4.0260</td><td> 3.0876</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENSG00000278814.1 </td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>⋯</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td><td>-9.9658</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>ENSG00000078237.5 </td><td> 3.5086</td><td> 2.3019</td><td> 0.0990</td><td> 3.6184</td><td> 3.5633</td><td> 2.9109</td><td> 2.1638</td><td> 2.6624</td><td> 1.2696</td><td>⋯</td><td> 1.5998</td><td> 2.8219</td><td> 2.5876</td><td> 2.6232</td><td> 2.9525</td><td> 3.0774</td><td> 3.4384</td><td> 2.2082</td><td> 2.1541</td><td>-0.5756</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 10536\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & sample & TCGA-19-1787-01 & TCGA-S9-A7J2-01 & TCGA-G3-A3CH-11 & TCGA-EK-A2RE-01 & TCGA-44-6778-01 & TCGA-F4-6854-01 & TCGA-AB-2863-03 & TCGA-C8-A1HL-01 & TCGA-EW-A2FS-01 & ⋯ & TCGA-DJ-A2QC-01 & TCGA-A8-A09K-01 & TCGA-61-1907-01 & TCGA-IB-7885-01 & TCGA-95-7947-01 & TCGA-VQ-AA6F-01 & TCGA-BR-8588-01 & TCGA-24-2254-01 & TCGA-DD-A115-01 & TCGA-FV-A3I0-11\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000242268.2  & -9.9658 &  0.2998 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & ⋯ & -1.3921 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -1.9379 & -9.9658 & -9.9658\\\\\n",
       "\t2 & ENSG00000259041.1  & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & ⋯ & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658\\\\\n",
       "\t3 & ENSG00000270112.3  & -3.8160 & -3.0469 & -9.9658 & -9.9658 & -5.5735 & -9.9658 & -3.4580 & -9.9658 & -5.0116 & ⋯ & -6.5064 & -9.9658 & -9.9658 & -9.9658 & -6.5064 & -9.9658 & -9.9658 & -5.5735 & -9.9658 & -9.9658\\\\\n",
       "\t4 & ENSG00000167578.16 &  5.2998 &  4.8881 &  3.5572 &  4.2563 &  5.3162 &  4.5161 &  3.6242 &  4.9782 &  5.7035 & ⋯ &  5.4591 &  5.5364 &  4.7798 &  5.0514 &  6.1607 &  5.2814 &  3.9599 &  5.0760 &  4.0260 &  3.0876\\\\\n",
       "\t5 & ENSG00000278814.1  & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & ⋯ & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658 & -9.9658\\\\\n",
       "\t6 & ENSG00000078237.5  &  3.5086 &  2.3019 &  0.0990 &  3.6184 &  3.5633 &  2.9109 &  2.1638 &  2.6624 &  1.2696 & ⋯ &  1.5998 &  2.8219 &  2.5876 &  2.6232 &  2.9525 &  3.0774 &  3.4384 &  2.2082 &  2.1541 & -0.5756\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 10536\n",
       "\n",
       "| <!--/--> | sample &lt;chr&gt; | TCGA-19-1787-01 &lt;dbl&gt; | TCGA-S9-A7J2-01 &lt;dbl&gt; | TCGA-G3-A3CH-11 &lt;dbl&gt; | TCGA-EK-A2RE-01 &lt;dbl&gt; | TCGA-44-6778-01 &lt;dbl&gt; | TCGA-F4-6854-01 &lt;dbl&gt; | TCGA-AB-2863-03 &lt;dbl&gt; | TCGA-C8-A1HL-01 &lt;dbl&gt; | TCGA-EW-A2FS-01 &lt;dbl&gt; | ⋯ ⋯ | TCGA-DJ-A2QC-01 &lt;dbl&gt; | TCGA-A8-A09K-01 &lt;dbl&gt; | TCGA-61-1907-01 &lt;dbl&gt; | TCGA-IB-7885-01 &lt;dbl&gt; | TCGA-95-7947-01 &lt;dbl&gt; | TCGA-VQ-AA6F-01 &lt;dbl&gt; | TCGA-BR-8588-01 &lt;dbl&gt; | TCGA-24-2254-01 &lt;dbl&gt; | TCGA-DD-A115-01 &lt;dbl&gt; | TCGA-FV-A3I0-11 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | ENSG00000242268.2  | -9.9658 |  0.2998 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | ⋯ | -1.3921 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -1.9379 | -9.9658 | -9.9658 |\n",
       "| 2 | ENSG00000259041.1  | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | ⋯ | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 |\n",
       "| 3 | ENSG00000270112.3  | -3.8160 | -3.0469 | -9.9658 | -9.9658 | -5.5735 | -9.9658 | -3.4580 | -9.9658 | -5.0116 | ⋯ | -6.5064 | -9.9658 | -9.9658 | -9.9658 | -6.5064 | -9.9658 | -9.9658 | -5.5735 | -9.9658 | -9.9658 |\n",
       "| 4 | ENSG00000167578.16 |  5.2998 |  4.8881 |  3.5572 |  4.2563 |  5.3162 |  4.5161 |  3.6242 |  4.9782 |  5.7035 | ⋯ |  5.4591 |  5.5364 |  4.7798 |  5.0514 |  6.1607 |  5.2814 |  3.9599 |  5.0760 |  4.0260 |  3.0876 |\n",
       "| 5 | ENSG00000278814.1  | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | ⋯ | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 | -9.9658 |\n",
       "| 6 | ENSG00000078237.5  |  3.5086 |  2.3019 |  0.0990 |  3.6184 |  3.5633 |  2.9109 |  2.1638 |  2.6624 |  1.2696 | ⋯ |  1.5998 |  2.8219 |  2.5876 |  2.6232 |  2.9525 |  3.0774 |  3.4384 |  2.2082 |  2.1541 | -0.5756 |\n",
       "\n"
      ],
      "text/plain": [
       "  sample             TCGA-19-1787-01 TCGA-S9-A7J2-01 TCGA-G3-A3CH-11\n",
       "1 ENSG00000242268.2  -9.9658          0.2998         -9.9658        \n",
       "2 ENSG00000259041.1  -9.9658         -9.9658         -9.9658        \n",
       "3 ENSG00000270112.3  -3.8160         -3.0469         -9.9658        \n",
       "4 ENSG00000167578.16  5.2998          4.8881          3.5572        \n",
       "5 ENSG00000278814.1  -9.9658         -9.9658         -9.9658        \n",
       "6 ENSG00000078237.5   3.5086          2.3019          0.0990        \n",
       "  TCGA-EK-A2RE-01 TCGA-44-6778-01 TCGA-F4-6854-01 TCGA-AB-2863-03\n",
       "1 -9.9658         -9.9658         -9.9658         -9.9658        \n",
       "2 -9.9658         -9.9658         -9.9658         -9.9658        \n",
       "3 -9.9658         -5.5735         -9.9658         -3.4580        \n",
       "4  4.2563          5.3162          4.5161          3.6242        \n",
       "5 -9.9658         -9.9658         -9.9658         -9.9658        \n",
       "6  3.6184          3.5633          2.9109          2.1638        \n",
       "  TCGA-C8-A1HL-01 TCGA-EW-A2FS-01 ⋯ TCGA-DJ-A2QC-01 TCGA-A8-A09K-01\n",
       "1 -9.9658         -9.9658         ⋯ -1.3921         -9.9658        \n",
       "2 -9.9658         -9.9658         ⋯ -9.9658         -9.9658        \n",
       "3 -9.9658         -5.0116         ⋯ -6.5064         -9.9658        \n",
       "4  4.9782          5.7035         ⋯  5.4591          5.5364        \n",
       "5 -9.9658         -9.9658         ⋯ -9.9658         -9.9658        \n",
       "6  2.6624          1.2696         ⋯  1.5998          2.8219        \n",
       "  TCGA-61-1907-01 TCGA-IB-7885-01 TCGA-95-7947-01 TCGA-VQ-AA6F-01\n",
       "1 -9.9658         -9.9658         -9.9658         -9.9658        \n",
       "2 -9.9658         -9.9658         -9.9658         -9.9658        \n",
       "3 -9.9658         -9.9658         -6.5064         -9.9658        \n",
       "4  4.7798          5.0514          6.1607          5.2814        \n",
       "5 -9.9658         -9.9658         -9.9658         -9.9658        \n",
       "6  2.5876          2.6232          2.9525          3.0774        \n",
       "  TCGA-BR-8588-01 TCGA-24-2254-01 TCGA-DD-A115-01 TCGA-FV-A3I0-11\n",
       "1 -9.9658         -1.9379         -9.9658         -9.9658        \n",
       "2 -9.9658         -9.9658         -9.9658         -9.9658        \n",
       "3 -9.9658         -5.5735         -9.9658         -9.9658        \n",
       "4  3.9599          5.0760          4.0260          3.0876        \n",
       "5 -9.9658         -9.9658         -9.9658         -9.9658        \n",
       "6  3.4384          2.2082          2.1541         -0.5756        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_tpm_copy)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e184ccce",
   "metadata": {},
   "source": [
    "# tcga query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0de9c1c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--------------------------------------\n",
      "\n",
      "o GDCquery: Searching in GDC database\n",
      "\n",
      "--------------------------------------\n",
      "\n",
      "Genome of reference: hg38\n",
      "\n",
      "--------------------------------------------\n",
      "\n",
      "oo Accessing GDC. This might take a while...\n",
      "\n",
      "--------------------------------------------\n",
      "\n",
      "ooo Project: TCGA-BRCA\n",
      "\n",
      "ooo Project: TCGA-MESO\n",
      "\n",
      "ooo Project: TCGA-CESC\n",
      "\n",
      "ooo Project: TCGA-BLCA\n",
      "\n",
      "ooo Project: TCGA-CHOL\n",
      "\n",
      "ooo Project: TCGA-ACC\n",
      "\n",
      "ooo Project: TCGA-DLBC\n",
      "\n",
      "ooo Project: TCGA-HNSC\n",
      "\n",
      "ooo Project: TCGA-COAD\n",
      "\n",
      "ooo Project: TCGA-ESCA\n",
      "\n",
      "ooo Project: TCGA-LAML\n",
      "\n",
      "ooo Project: TCGA-KIRP\n",
      "\n",
      "ooo Project: TCGA-KIRC\n",
      "\n",
      "ooo Project: TCGA-GBM\n",
      "\n",
      "ooo Project: TCGA-KICH\n",
      "\n",
      "ooo Project: TCGA-READ\n",
      "\n",
      "ooo Project: TCGA-PAAD\n",
      "\n",
      "ooo Project: TCGA-LUAD\n",
      "\n",
      "ooo Project: TCGA-OV\n",
      "\n",
      "ooo Project: TCGA-LIHC\n",
      "\n",
      "ooo Project: TCGA-LUSC\n",
      "\n",
      "ooo Project: TCGA-LGG\n",
      "\n",
      "ooo Project: TCGA-UCEC\n",
      "\n",
      "ooo Project: TCGA-PRAD\n",
      "\n",
      "ooo Project: TCGA-STAD\n",
      "\n",
      "ooo Project: TCGA-THCA\n",
      "\n",
      "ooo Project: TCGA-THYM\n",
      "\n",
      "ooo Project: TCGA-UCS\n",
      "\n",
      "ooo Project: TCGA-UVM\n",
      "\n",
      "ooo Project: TCGA-PCPG\n",
      "\n",
      "ooo Project: TCGA-SARC\n",
      "\n",
      "ooo Project: TCGA-SKCM\n",
      "\n",
      "ooo Project: TCGA-TGCT\n",
      "\n",
      "--------------------\n",
      "\n",
      "oo Filtering results\n",
      "\n",
      "--------------------\n",
      "\n",
      "ooo By data.type\n",
      "\n",
      "----------------\n",
      "\n",
      "oo Checking data\n",
      "\n",
      "----------------\n",
      "\n",
      "ooo Check if there are duplicated cases\n",
      "\n",
      "Warning: There are more than one file for the same case. Please verify query results. You can use the command View(getResults(query)) in rstudio\n",
      "\n",
      "ooo Check if there results for the query\n",
      "\n",
      "-------------------\n",
      "\n",
      "o Preparing output\n",
      "\n",
      "-------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "tcga_projects <- TCGAbiolinks:::getGDCprojects()$project_id[grep('TCGA',TCGAbiolinks:::getGDCprojects()$project_id)]\n",
    "query<-GDCquery(tcga_projects,\n",
    "         data.category = \"Transcriptome Profiling\",\n",
    "         data.type = \"Gene Expression Quantification\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5e86062e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'TCGA-BRCA'</li><li>'TCGA-MESO'</li><li>'TCGA-CESC'</li><li>'TCGA-BLCA'</li><li>'TCGA-CHOL'</li><li>'TCGA-ACC'</li><li>'TCGA-DLBC'</li><li>'TCGA-HNSC'</li><li>'TCGA-COAD'</li><li>'TCGA-ESCA'</li><li>'TCGA-LAML'</li><li>'TCGA-KIRP'</li><li>'TCGA-KIRC'</li><li>'TCGA-GBM'</li><li>'TCGA-KICH'</li><li>'TCGA-READ'</li><li>'TCGA-PAAD'</li><li>'TCGA-LUAD'</li><li>'TCGA-OV'</li><li>'TCGA-LIHC'</li><li>'TCGA-LUSC'</li><li>'TCGA-LGG'</li><li>'TCGA-UCEC'</li><li>'TCGA-PRAD'</li><li>'TCGA-STAD'</li><li>'TCGA-THCA'</li><li>'TCGA-THYM'</li><li>'TCGA-UCS'</li><li>'TCGA-UVM'</li><li>'TCGA-PCPG'</li><li>'TCGA-SARC'</li><li>'TCGA-SKCM'</li><li>'TCGA-TGCT'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'TCGA-BRCA'\n",
       "\\item 'TCGA-MESO'\n",
       "\\item 'TCGA-CESC'\n",
       "\\item 'TCGA-BLCA'\n",
       "\\item 'TCGA-CHOL'\n",
       "\\item 'TCGA-ACC'\n",
       "\\item 'TCGA-DLBC'\n",
       "\\item 'TCGA-HNSC'\n",
       "\\item 'TCGA-COAD'\n",
       "\\item 'TCGA-ESCA'\n",
       "\\item 'TCGA-LAML'\n",
       "\\item 'TCGA-KIRP'\n",
       "\\item 'TCGA-KIRC'\n",
       "\\item 'TCGA-GBM'\n",
       "\\item 'TCGA-KICH'\n",
       "\\item 'TCGA-READ'\n",
       "\\item 'TCGA-PAAD'\n",
       "\\item 'TCGA-LUAD'\n",
       "\\item 'TCGA-OV'\n",
       "\\item 'TCGA-LIHC'\n",
       "\\item 'TCGA-LUSC'\n",
       "\\item 'TCGA-LGG'\n",
       "\\item 'TCGA-UCEC'\n",
       "\\item 'TCGA-PRAD'\n",
       "\\item 'TCGA-STAD'\n",
       "\\item 'TCGA-THCA'\n",
       "\\item 'TCGA-THYM'\n",
       "\\item 'TCGA-UCS'\n",
       "\\item 'TCGA-UVM'\n",
       "\\item 'TCGA-PCPG'\n",
       "\\item 'TCGA-SARC'\n",
       "\\item 'TCGA-SKCM'\n",
       "\\item 'TCGA-TGCT'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'TCGA-BRCA'\n",
       "2. 'TCGA-MESO'\n",
       "3. 'TCGA-CESC'\n",
       "4. 'TCGA-BLCA'\n",
       "5. 'TCGA-CHOL'\n",
       "6. 'TCGA-ACC'\n",
       "7. 'TCGA-DLBC'\n",
       "8. 'TCGA-HNSC'\n",
       "9. 'TCGA-COAD'\n",
       "10. 'TCGA-ESCA'\n",
       "11. 'TCGA-LAML'\n",
       "12. 'TCGA-KIRP'\n",
       "13. 'TCGA-KIRC'\n",
       "14. 'TCGA-GBM'\n",
       "15. 'TCGA-KICH'\n",
       "16. 'TCGA-READ'\n",
       "17. 'TCGA-PAAD'\n",
       "18. 'TCGA-LUAD'\n",
       "19. 'TCGA-OV'\n",
       "20. 'TCGA-LIHC'\n",
       "21. 'TCGA-LUSC'\n",
       "22. 'TCGA-LGG'\n",
       "23. 'TCGA-UCEC'\n",
       "24. 'TCGA-PRAD'\n",
       "25. 'TCGA-STAD'\n",
       "26. 'TCGA-THCA'\n",
       "27. 'TCGA-THYM'\n",
       "28. 'TCGA-UCS'\n",
       "29. 'TCGA-UVM'\n",
       "30. 'TCGA-PCPG'\n",
       "31. 'TCGA-SARC'\n",
       "32. 'TCGA-SKCM'\n",
       "33. 'TCGA-TGCT'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"TCGA-BRCA\" \"TCGA-MESO\" \"TCGA-CESC\" \"TCGA-BLCA\" \"TCGA-CHOL\" \"TCGA-ACC\" \n",
       " [7] \"TCGA-DLBC\" \"TCGA-HNSC\" \"TCGA-COAD\" \"TCGA-ESCA\" \"TCGA-LAML\" \"TCGA-KIRP\"\n",
       "[13] \"TCGA-KIRC\" \"TCGA-GBM\"  \"TCGA-KICH\" \"TCGA-READ\" \"TCGA-PAAD\" \"TCGA-LUAD\"\n",
       "[19] \"TCGA-OV\"   \"TCGA-LIHC\" \"TCGA-LUSC\" \"TCGA-LGG\"  \"TCGA-UCEC\" \"TCGA-PRAD\"\n",
       "[25] \"TCGA-STAD\" \"TCGA-THCA\" \"TCGA-THYM\" \"TCGA-UCS\"  \"TCGA-UVM\"  \"TCGA-PCPG\"\n",
       "[31] \"TCGA-SARC\" \"TCGA-SKCM\" \"TCGA-TGCT\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "TCGAbiolinks:::getGDCprojects()$project_id[grep('TCGA',TCGAbiolinks:::getGDCprojects()$project_id)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9fab84c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "queryDF<- getResults(query)\n",
    "project_types <- character(length(gene_tpm_copy)-1)\n",
    "\n",
    "for(i in tpm_length) {\n",
    "    sample <- colnames(gene_tpm_copy)[i]\n",
    "    sample <- substr(sample,1,12)\n",
    "    project_types[i-1] <- queryDF[queryDF$cases.submitter_id ==  sample,]$project[1]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cabbe0a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".dl-inline {width: auto; margin:0; padding: 0}\n",
       ".dl-inline>dt, .dl-inline>dd {float: none; width: auto; display: inline-block}\n",
       ".dl-inline>dt::after {content: \":\\0020\"; padding-right: .5ex}\n",
       ".dl-inline>dt:not(:first-of-type) {padding-left: .5ex}\n",
       "</style><dl class=dl-inline><dt>TCGA-ACC</dt><dd>77</dd><dt>TCGA-BLCA</dt><dd>426</dd><dt>TCGA-BRCA</dt><dd>1209</dd><dt>TCGA-CESC</dt><dd>309</dd><dt>TCGA-CHOL</dt><dd>45</dd><dt>TCGA-COAD</dt><dd>331</dd><dt>TCGA-DLBC</dt><dd>47</dd><dt>TCGA-ESCA</dt><dd>175</dd><dt>TCGA-GBM</dt><dd>170</dd><dt>TCGA-HNSC</dt><dd>544</dd><dt>TCGA-KICH</dt><dd>91</dd><dt>TCGA-KIRC</dt><dd>600</dd><dt>TCGA-KIRP</dt><dd>320</dd><dt>TCGA-LAML</dt><dd>147</dd><dt>TCGA-LGG</dt><dd>520</dd><dt>TCGA-LIHC</dt><dd>421</dd><dt>TCGA-LUAD</dt><dd>572</dd><dt>TCGA-LUSC</dt><dd>548</dd><dt>TCGA-MESO</dt><dd>86</dd><dt>TCGA-OV</dt><dd>380</dd><dt>TCGA-PAAD</dt><dd>182</dd><dt>TCGA-PCPG</dt><dd>185</dd><dt>TCGA-PRAD</dt><dd>546</dd><dt>TCGA-READ</dt><dd>103</dd><dt>TCGA-SARC</dt><dd>264</dd><dt>TCGA-SKCM</dt><dd>469</dd><dt>TCGA-STAD</dt><dd>411</dd><dt>TCGA-TGCT</dt><dd>154</dd><dt>TCGA-THCA</dt><dd>568</dd><dt>TCGA-THYM</dt><dd>120</dd><dt>TCGA-UCEC</dt><dd>202</dd><dt>TCGA-UCS</dt><dd>56</dd><dt>TCGA-UVM</dt><dd>79</dd><dt>NA's</dt><dd>178</dd></dl>\n"
      ],
      "text/latex": [
       "\\begin{description*}\n",
       "\\item[TCGA-ACC] 77\n",
       "\\item[TCGA-BLCA] 426\n",
       "\\item[TCGA-BRCA] 1209\n",
       "\\item[TCGA-CESC] 309\n",
       "\\item[TCGA-CHOL] 45\n",
       "\\item[TCGA-COAD] 331\n",
       "\\item[TCGA-DLBC] 47\n",
       "\\item[TCGA-ESCA] 175\n",
       "\\item[TCGA-GBM] 170\n",
       "\\item[TCGA-HNSC] 544\n",
       "\\item[TCGA-KICH] 91\n",
       "\\item[TCGA-KIRC] 600\n",
       "\\item[TCGA-KIRP] 320\n",
       "\\item[TCGA-LAML] 147\n",
       "\\item[TCGA-LGG] 520\n",
       "\\item[TCGA-LIHC] 421\n",
       "\\item[TCGA-LUAD] 572\n",
       "\\item[TCGA-LUSC] 548\n",
       "\\item[TCGA-MESO] 86\n",
       "\\item[TCGA-OV] 380\n",
       "\\item[TCGA-PAAD] 182\n",
       "\\item[TCGA-PCPG] 185\n",
       "\\item[TCGA-PRAD] 546\n",
       "\\item[TCGA-READ] 103\n",
       "\\item[TCGA-SARC] 264\n",
       "\\item[TCGA-SKCM] 469\n",
       "\\item[TCGA-STAD] 411\n",
       "\\item[TCGA-TGCT] 154\n",
       "\\item[TCGA-THCA] 568\n",
       "\\item[TCGA-THYM] 120\n",
       "\\item[TCGA-UCEC] 202\n",
       "\\item[TCGA-UCS] 56\n",
       "\\item[TCGA-UVM] 79\n",
       "\\item[NA's] 178\n",
       "\\end{description*}\n"
      ],
      "text/markdown": [
       "TCGA-ACC\n",
       ":   77TCGA-BLCA\n",
       ":   426TCGA-BRCA\n",
       ":   1209TCGA-CESC\n",
       ":   309TCGA-CHOL\n",
       ":   45TCGA-COAD\n",
       ":   331TCGA-DLBC\n",
       ":   47TCGA-ESCA\n",
       ":   175TCGA-GBM\n",
       ":   170TCGA-HNSC\n",
       ":   544TCGA-KICH\n",
       ":   91TCGA-KIRC\n",
       ":   600TCGA-KIRP\n",
       ":   320TCGA-LAML\n",
       ":   147TCGA-LGG\n",
       ":   520TCGA-LIHC\n",
       ":   421TCGA-LUAD\n",
       ":   572TCGA-LUSC\n",
       ":   548TCGA-MESO\n",
       ":   86TCGA-OV\n",
       ":   380TCGA-PAAD\n",
       ":   182TCGA-PCPG\n",
       ":   185TCGA-PRAD\n",
       ":   546TCGA-READ\n",
       ":   103TCGA-SARC\n",
       ":   264TCGA-SKCM\n",
       ":   469TCGA-STAD\n",
       ":   411TCGA-TGCT\n",
       ":   154TCGA-THCA\n",
       ":   568TCGA-THYM\n",
       ":   120TCGA-UCEC\n",
       ":   202TCGA-UCS\n",
       ":   56TCGA-UVM\n",
       ":   79NA's\n",
       ":   178\n",
       "\n"
      ],
      "text/plain": [
       " TCGA-ACC TCGA-BLCA TCGA-BRCA TCGA-CESC TCGA-CHOL TCGA-COAD TCGA-DLBC TCGA-ESCA \n",
       "       77       426      1209       309        45       331        47       175 \n",
       " TCGA-GBM TCGA-HNSC TCGA-KICH TCGA-KIRC TCGA-KIRP TCGA-LAML  TCGA-LGG TCGA-LIHC \n",
       "      170       544        91       600       320       147       520       421 \n",
       "TCGA-LUAD TCGA-LUSC TCGA-MESO   TCGA-OV TCGA-PAAD TCGA-PCPG TCGA-PRAD TCGA-READ \n",
       "      572       548        86       380       182       185       546       103 \n",
       "TCGA-SARC TCGA-SKCM TCGA-STAD TCGA-TGCT TCGA-THCA TCGA-THYM TCGA-UCEC  TCGA-UCS \n",
       "      264       469       411       154       568       120       202        56 \n",
       " TCGA-UVM      NA's \n",
       "       79       178 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(as.factor(project_types))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "88b4d421",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'10536'"
      ],
      "text/latex": [
       "'10536'"
      ],
      "text/markdown": [
       "'10536'"
      ],
      "text/plain": [
       "[1] \"10536\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rownames(gene_tpm_copy)[i]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "491549b8",
   "metadata": {},
   "source": [
    "# functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f7f12528",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### functions\n",
    "# function to retrieve the mean zscore for a tissue type\n",
    "get.most.expressed <- function(tissue_type) {\n",
    "    # calculate mean z_score\n",
    "    soi <- colnames(gene_tpm_copy[,tpm_length])[project_types == tissue_type & !is.na(project_types)]\n",
    "\n",
    "    temp_samples <- gene_tpm_copy[,tpm_length][,soi]\n",
    "    \n",
    "    mean_zscore <- rowMeans(temp_samples, na.rm = F)\n",
    "    \n",
    "    # extract the columns for the tissues of interest, in this case we want Lung, Liver, Esophagus, Bladder, and Intestine\n",
    "    tissue_mean_zscore <- data.frame(cbind(\"gene\" = gene_tpm_copy[,1], \"mean_Zscore\" = mean_zscore))\n",
    "    \n",
    "    # select the top 1000 most highly expressed genes\n",
    "    tissue_most_expressed <- tissue_mean_zscore[with(tissue_mean_zscore, order(mean_Zscore, decreasing = T)),]\n",
    "    print(head(tissue_most_expressed))\n",
    "    tissue_most_expressed <- tissue_most_expressed[1:1000,]\n",
    "    \n",
    "    current_hgnc <- getLDS(attributes=c(\"ensembl_gene_id\"), filters=\"ensembl_gene_id\",\n",
    "                             values=gsub(\"\\\\..*\",\"\", tissue_most_expressed$gene),\n",
    "                             mart=human,attributesL=c(\"hgnc_symbol\"),\n",
    "                             martL=human)\n",
    "    \n",
    "    return(current_hgnc$HGNC.symbol)\n",
    "}\n",
    "\n",
    "compute_enrichment <- function(foreground_genes,all_genes,background_genes=NULL,pathways=NULL) {\n",
    "    if (is.null(pathways)) { \n",
    "        pathways <- load_pathways()\n",
    "        pathways <- lapply( pathways, function(pathway_genes) {return(pathway_genes[pathway_genes %in% all_genes])})\n",
    "    }\n",
    "    fisher_enrichment_dt <- data.table(pathway=names(pathways),p_value=-1,odds_ratio=-1)\n",
    "    if (is.null(background_genes)) {\n",
    "        background_genes <- setdiff(all_genes,foreground_genes)\n",
    "        #background_genes <- unique(unlist(pathways))\n",
    "    }\n",
    "\n",
    "    for (pathway in names(pathways)) {\n",
    "        pathway_genes <- pathways[[pathway]]\n",
    "        non_pathway_genes <- setdiff(all_genes,pathway_genes)\n",
    "\n",
    "        num_in_pathway_and_foreground <- intersect(pathway_genes,foreground_genes) %>% length\n",
    "        num_in_pathway_and_not_foreground <- intersect(pathway_genes,background_genes) %>% length\n",
    "        num_not_in_pathway_and_foreground <- intersect(non_pathway_genes,foreground_genes) %>% length\n",
    "        num_not_in_pathway_and_not_foreground <- intersect(non_pathway_genes,background_genes) %>% length\n",
    "        fisher_mat <- matrix(c(num_in_pathway_and_foreground,num_in_pathway_and_not_foreground,\n",
    "                              num_not_in_pathway_and_foreground,num_not_in_pathway_and_not_foreground),\n",
    "                             nrow=2,ncol=2,byrow=T)\n",
    "        \n",
    "        test_res <- fisher.test(fisher_mat,alternative=\"g\")\n",
    "        pathway_ <- pathway\n",
    "        fisher_enrichment_dt[pathway==pathway_,`:=`(p_value=test_res$p.value, odds_ratio=test_res$estimate,\n",
    "        num_p_fg=num_in_pathway_and_foreground,num_p_bg=num_in_pathway_and_not_foreground,\n",
    "        num_not_p_fg=num_not_in_pathway_and_foreground,num_not_p_bg=num_not_in_pathway_and_not_foreground)]\n",
    "    }\n",
    "    fisher_enrichment_dt[,q_value:=p.adjust(p_value)]\n",
    "\n",
    "    return(fisher_enrichment_dt)\n",
    "}\n",
    "\n",
    "get_enrichment_data <- function(current_files, current_pathways){\n",
    "    total_enrichment_pathways <- names(current_pathways)\n",
    "    pathwaysDF <- data.frame(\"pathways\" = total_enrichment_pathways)\n",
    "    pathwaysDF_odds <- data.frame(\"pathways\" = total_enrichment_pathways)\n",
    "    for(i in 1:length(current_files)) {\n",
    "        spec_exptype <- names(current_files)[i]\n",
    "        \n",
    "        current_hgnc <- getLDS(attributes=c(\"entrezgene_id\"), filters=\"entrezgene_id\",\n",
    "                             values=current_files[[spec_exptype]],\n",
    "                             mart=human,attributesL=c(\"hgnc_symbol\"),\n",
    "                             martL=human)\n",
    "        total_genes <- current_hgnc$HGNC.symbol\n",
    "        \n",
    "        \n",
    "        enrichment_test <- compute_enrichment(foreground_genes = total_genes,\n",
    "                                               all_genes = background_set$gene,\n",
    "                                               pathways = current_pathways)\n",
    "        subset_enrichment_test  <- enrichment_test[,c(\"q_value\")]\n",
    "        subset_enrichment_test_odds  <- enrichment_test[,c(\"odds_ratio\")]\n",
    "        rownames(subset_enrichment_test) <-  enrichment_test$pathway\n",
    "        rownames(subset_enrichment_test_odds) <-  enrichment_test$pathway\n",
    "\n",
    "        pathwaysDF <- cbind(pathwaysDF, subset_enrichment_test)\n",
    "        pathwaysDF_odds <- cbind(pathwaysDF_odds, subset_enrichment_test_odds)\n",
    "        #saveRDS(subset_enrichment_test, file = new_file_name)\n",
    "    }\n",
    "    \n",
    "    all_qvals <- pathwaysDF[,2:length(pathwaysDF)]\n",
    "    rownames(all_qvals) <- total_enrichment_pathways\n",
    "    colnames(all_qvals) <- names(current_files)\n",
    "    \n",
    "    all_odds <- pathwaysDF_odds[,2:length(pathwaysDF_odds)]\n",
    "    rownames(all_odds) <- total_enrichment_pathways\n",
    "    colnames(all_odds) <- names(current_files)\n",
    "    \n",
    "    enrichment_heatmap <- pheatmap(as.matrix(log(all_qvals)), fontsize = 8)\n",
    "    return(list(\"matrix\" = all_qvals, \"odds_matrix\" = all_odds, \"heatmap\" = enrichment_heatmap))\n",
    "}\n",
    "makeColorRampPalette <- function(colors, cutoff.fraction, num.colors.in.palette)\n",
    "{\n",
    "  stopifnot(length(colors) == 4)\n",
    "  ramp1 <- colorRampPalette(colors[1:2])(num.colors.in.palette * cutoff.fraction)\n",
    "  ramp2 <- colorRampPalette(colors[3:4])(num.colors.in.palette * (1 - cutoff.fraction))\n",
    "  return(c(ramp1, ramp2))\n",
    "}\n",
    "change_matrix <- function(fdr_mat, odds_mat, threshhold) {\n",
    "    new_mat <- matrix(nrow = nrow(fdr_mat), ncol = ncol(fdr_mat))\n",
    "    \n",
    "    for(i in 1:nrow(fdr_mat)) {\n",
    "        for(j in 1:ncol(fdr_mat)) {\n",
    "            \n",
    "            if(fdr_mat[i,j] < .05) {\n",
    "                new_mat[i,j] <- odds_mat[i,j]\n",
    "            }else {\n",
    "                new_mat[i,j] <- 0.001\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    rownames(new_mat) <- rownames(fdr_mat)\n",
    "    colnames(new_mat) <- colnames(fdr_mat)\n",
    "    \n",
    "    cutoff.distance <- 0.002  \n",
    "    cols <- makeColorRampPalette(c(\"white\", \"white\",    # distances 0 to 0.002 colored from white to red\n",
    "                                   \"red\", \"blue\"), # distances 0.002 to max(distmat) colored from green to black\n",
    "                                 cutoff.distance / max(new_mat),\n",
    "                                 100)\n",
    "    freq_mat <- apply(new_mat, 1, function(x) {sum(x > threshhold)/nrow(new_mat)})\n",
    "    freq_dat <- data.frame(\"cancertype\" = names(freq_mat),\n",
    "                     \"fraction\" = freq_mat)\n",
    "    \n",
    "    # Barplot\n",
    "    bar_plot<- ggplot(freq_dat, aes(x=cancertype, y=fraction)) + \n",
    "    geom_bar(stat = \"identity\") +\n",
    "    coord_flip()\n",
    "\n",
    "    new_heatmap <- pheatmap(new_mat,\n",
    "                            color = cols,\n",
    "                            fontsize = 8)\n",
    "    \n",
    "    return(list(\"matrix\"=new_mat, \"heatmap\" = new_heatmap, \"barplot\" = bar_plot))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa360021",
   "metadata": {},
   "source": [
    "# top 1000 project specific genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f0a7972b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    gene      mean_Zscore\n",
      "43217  ENSG00000237550.5 9.98670019265914\n",
      "41096 ENSG00000135404.11  9.9574028825187\n",
      "51704 ENSG00000167658.15 9.93701361989129\n",
      "33801 ENSG00000100219.16  9.9286447847754\n",
      "56772 ENSG00000135486.17 9.91615700196278\n",
      "57481  ENSG00000163041.9 9.91544389706719\n",
      "                    gene      mean_Zscore\n",
      "25793 ENSG00000070756.13  9.9820023255814\n",
      "23303  ENSG00000166794.4 9.97715697674419\n",
      "39006 ENSG00000172270.18 9.91845930232558\n",
      "49811  ENSG00000220842.6 9.90539651162791\n",
      "53177 ENSG00000104529.17 9.90226279069767\n",
      "3208  ENSG00000185201.16 9.88411395348837\n",
      "                    gene      mean_Zscore\n",
      "18879 ENSG00000170315.13 9.99025177993527\n",
      "57481  ENSG00000163041.9 9.98092556634304\n",
      "16025 ENSG00000067225.17 9.97515372168285\n",
      "1938  ENSG00000163682.15 9.97483980582524\n",
      "40962 ENSG00000123416.15 9.95027993527508\n",
      "56772 ENSG00000135486.17 9.94450938511327\n",
      "                    gene      mean_Zscore\n",
      "10313  ENSG00000234745.9 9.91515414740577\n",
      "2795  ENSG00000074800.13 9.89258945590488\n",
      "23850  ENSG00000110700.6 9.87972984828979\n",
      "57499  ENSG00000132475.8 9.87939505504178\n",
      "15600  ENSG00000233927.4  9.8764614497362\n",
      "56772 ENSG00000135486.17 9.85986950287338\n",
      "                    gene      mean_Zscore\n",
      "50356 ENSG00000019582.14 9.98700222222222\n",
      "40697 ENSG00000162244.10 9.97888888888889\n",
      "10611 ENSG00000140264.19 9.95519777777778\n",
      "19294 ENSG00000115414.18 9.94067333333333\n",
      "12    ENSG00000198242.13 9.89661333333333\n",
      "23850  ENSG00000110700.6 9.88542222222222\n",
      "                    gene      mean_Zscore\n",
      "59604  ENSG00000221983.7 9.96519480519481\n",
      "45379 ENSG00000182899.14 9.95091688311688\n",
      "28492 ENSG00000142676.12 9.90617792207792\n",
      "20720  ENSG00000125148.6 9.90455714285714\n",
      "52303  ENSG00000197728.9 9.88593506493507\n",
      "60453 ENSG00000102265.11  9.8797974025974\n",
      "                    gene      mean_Zscore\n",
      "29800  ENSG00000169442.8 9.95442127659575\n",
      "27781  ENSG00000131143.8 9.93666382978723\n",
      "22949  ENSG00000219928.2 9.91868723404255\n",
      "45682 ENSG00000102879.15 9.91805531914894\n",
      "2250  ENSG00000132341.11 9.89306382978723\n",
      "50988 ENSG00000123349.13 9.87219574468085\n",
      "                    gene      mean_Zscore\n",
      "12    ENSG00000198242.13 9.99748602941176\n",
      "58256 ENSG00000174748.18 9.97848878676471\n",
      "18879 ENSG00000170315.13 9.97253584558823\n",
      "5847  ENSG00000149806.10 9.96250643382353\n",
      "23850  ENSG00000110700.6 9.95339816176471\n",
      "535   ENSG00000134333.13 9.91142867647059\n",
      "                    gene      mean_Zscore\n",
      "10611 ENSG00000140264.19 9.99346509757196\n",
      "36775 ENSG00000206503.11 9.98002236411251\n",
      "15600  ENSG00000233927.4  9.9648632798387\n",
      "7253  ENSG00000119888.10 9.93894708151958\n",
      "22949  ENSG00000219928.2 9.88573733593734\n",
      "24298 ENSG00000114942.13 9.87010217183054\n",
      "                    gene      mean_Zscore\n",
      "21556 ENSG00000197958.12 9.99940628571429\n",
      "58646 ENSG00000136156.12         9.978228\n",
      "22370  ENSG00000276168.1 9.97744342857143\n",
      "52167  ENSG00000279483.1 9.97477371428571\n",
      "33974 ENSG00000171863.12 9.97311142857143\n",
      "22710 ENSG00000182718.16 9.95435314285714\n",
      "                    gene      mean_Zscore\n",
      "45379 ENSG00000182899.14  9.9859462585034\n",
      "56572  ENSG00000263934.4 9.95811836734694\n",
      "18879 ENSG00000170315.13 9.94625714285714\n",
      "25645 ENSG00000181163.13 9.93747755102041\n",
      "33974 ENSG00000171863.12  9.9192843537415\n",
      "51704 ENSG00000167658.15 9.91697823129252\n",
      "                    gene  mean_Zscore\n",
      "33974 ENSG00000171863.12    9.9836625\n",
      "13425  ENSG00000112306.7 9.9597428125\n",
      "14255 ENSG00000150991.14  9.952349375\n",
      "51704 ENSG00000167658.15  9.950653125\n",
      "27781  ENSG00000131143.8 9.9421253125\n",
      "26369 ENSG00000163453.11      9.94111\n",
      "                    gene      mean_Zscore\n",
      "59046  ENSG00000272398.5 9.99899969269895\n",
      "48423 ENSG00000204525.14  9.9953275477254\n",
      "51704 ENSG00000167658.15  9.9657124062286\n",
      "23922 ENSG00000130255.12 9.95029066622588\n",
      "44413 ENSG00000187514.14 9.93955127380877\n",
      "20227 ENSG00000117984.12 9.93507547330442\n",
      "                    gene      mean_Zscore\n",
      "43217  ENSG00000237550.5 9.99388042496047\n",
      "23072 ENSG00000104904.12 9.97091619755176\n",
      "50222 ENSG00000182774.10    9.96504790814\n",
      "23922 ENSG00000130255.12 9.95674073897635\n",
      "7039   ENSG00000274012.1 9.95142439884006\n",
      "20720  ENSG00000125148.6 9.94561221745588\n",
      "                    gene      mean_Zscore\n",
      "53899 ENSG00000114391.12 9.99727472527472\n",
      "48329 ENSG00000166165.12 9.98912197802198\n",
      "39108  ENSG00000108298.9 9.98563076923077\n",
      "7039   ENSG00000274012.1 9.94602747252747\n",
      "7183  ENSG00000134419.15 9.94310879120879\n",
      "51704 ENSG00000167658.15 9.93075494505494\n",
      "                    gene      mean_Zscore\n",
      "24298 ENSG00000114942.13 9.99413980582524\n",
      "10611 ENSG00000140264.19 9.98474368932039\n",
      "24685  ENSG00000275215.1  9.9771572815534\n",
      "26276  ENSG00000277739.1  9.9771572815534\n",
      "30491  ENSG00000278233.1  9.9771572815534\n",
      "31908  ENSG00000278189.1  9.9771572815534\n",
      "                    gene      mean_Zscore\n",
      "33974 ENSG00000171863.12 9.99871483516483\n",
      "3546  ENSG00000188846.13 9.99060714285714\n",
      "25240 ENSG00000122406.12 9.95953791208791\n",
      "26330 ENSG00000198034.10 9.95189450549451\n",
      "50812 ENSG00000108821.13 9.94853351648352\n",
      "56979  ENSG00000083845.8 9.94252032967033\n",
      "                    gene      mean_Zscore\n",
      "50126 ENSG00000118181.10  9.9957972971001\n",
      "14255 ENSG00000150991.14 9.98831348212675\n",
      "56979  ENSG00000083845.8 9.98466714877185\n",
      "23850  ENSG00000110700.6 9.98125282983509\n",
      "25240 ENSG00000122406.12 9.97702749686872\n",
      "5847  ENSG00000149806.10 9.91904286377503\n",
      "                    gene      mean_Zscore\n",
      "22069 ENSG00000160932.10 9.98306566199053\n",
      "40962 ENSG00000123416.15 9.95707189748263\n",
      "16563 ENSG00000196531.10 9.93257646233776\n",
      "38829 ENSG00000110492.15 9.93225398075474\n",
      "3546  ENSG00000188846.13 9.91783029878421\n",
      "5847  ENSG00000149806.10 9.87127218181934\n",
      "                    gene      mean_Zscore\n",
      "28492 ENSG00000142676.12 9.99715569999967\n",
      "37490 ENSG00000109475.16 9.99351440487981\n",
      "49137 ENSG00000110169.10 9.97560822696917\n",
      "13425  ENSG00000112306.7 9.96799291106841\n",
      "39374  ENSG00000135744.7 9.96620654804311\n",
      "40697 ENSG00000162244.10  9.9499995707342\n",
      "                    gene      mean_Zscore\n",
      "1938  ENSG00000163682.15 9.99597062218996\n",
      "25240 ENSG00000122406.12   9.987539952414\n",
      "57499  ENSG00000132475.8 9.97817613429664\n",
      "26330 ENSG00000198034.10 9.94974483601642\n",
      "38996 ENSG00000172757.12  9.9460600573427\n",
      "10313  ENSG00000234745.9 9.94351003627774\n",
      "                    gene      mean_Zscore\n",
      "11771  ENSG00000177600.8 9.98910481649377\n",
      "23850  ENSG00000110700.6 9.98596691996648\n",
      "4082   ENSG00000234851.4  9.9838862292215\n",
      "3854  ENSG00000111716.12 9.95087891841452\n",
      "56772 ENSG00000135486.17 9.94668636483458\n",
      "33419  ENSG00000241343.9 9.94388925616125\n",
      "                    gene      mean_Zscore\n",
      "20636 ENSG00000179218.13 9.99991892548832\n",
      "2795  ENSG00000074800.13 9.93496362902846\n",
      "14255 ENSG00000150991.14 9.90146025402243\n",
      "38829 ENSG00000110492.15 9.88762876696713\n",
      "23303  ENSG00000166794.4 9.85977669053653\n",
      "55525 ENSG00000142089.15  9.8551169182947\n",
      "                    gene      mean_Zscore\n",
      "53177 ENSG00000104529.17 9.96165247787985\n",
      "49811  ENSG00000220842.6 9.94697489800308\n",
      "14935 ENSG00000122026.10  9.8697506345194\n",
      "48329 ENSG00000166165.12 9.83726591773835\n",
      "44413 ENSG00000187514.14 9.80015981314346\n",
      "23303  ENSG00000166794.4 9.79593863835368\n",
      "                    gene      mean_Zscore\n",
      "35706 ENSG00000198755.10 9.99614695863747\n",
      "22752 ENSG00000071082.10 9.99293187347932\n",
      "12    ENSG00000198242.13 9.99279927007299\n",
      "728   ENSG00000149925.16 9.98856690997567\n",
      "10313  ENSG00000234745.9  9.9529498783455\n",
      "31400  ENSG00000125534.9 9.93059562043796\n",
      "                    gene      mean_Zscore\n",
      "15600  ENSG00000233927.4 9.99890299295775\n",
      "20053 ENSG00000130066.16 9.98156179577465\n",
      "50988 ENSG00000123349.13 9.97443257042254\n",
      "10313  ENSG00000234745.9  9.9678588028169\n",
      "56979  ENSG00000083845.8 9.95131672535211\n",
      "59046  ENSG00000272398.5 9.94876461267606\n",
      "                    gene      mean_Zscore\n",
      "27781  ENSG00000131143.8 9.97882583333333\n",
      "5902  ENSG00000127184.10 9.97144833333333\n",
      "29399  ENSG00000211772.8 9.97135916666667\n",
      "22768 ENSG00000087460.23 9.94570916666667\n",
      "38996 ENSG00000172757.12 9.94568333333333\n",
      "16435 ENSG00000145741.15 9.90536333333333\n",
      "                    gene      mean_Zscore\n",
      "2795  ENSG00000074800.13 9.98444642857143\n",
      "38996 ENSG00000172757.12 9.97338214285714\n",
      "32855  ENSG00000229119.3 9.93611428571429\n",
      "1893   ENSG00000196656.7 9.91189107142857\n",
      "35995 ENSG00000100097.11         9.909675\n",
      "52181  ENSG00000106211.8 9.90941785714286\n",
      "                    gene      mean_Zscore\n",
      "36775 ENSG00000206503.11 9.98980506329114\n",
      "6665   ENSG00000170296.9  9.8815253164557\n",
      "56772 ENSG00000135486.17 9.84990632911392\n",
      "52181  ENSG00000106211.8 9.83873291139241\n",
      "18970 ENSG00000169100.12 9.79657215189873\n",
      "27781  ENSG00000131143.8 9.79173924050633\n",
      "                    gene      mean_Zscore\n",
      "51163  ENSG00000034510.5 9.96574162162162\n",
      "7039   ENSG00000274012.1 9.95421891891892\n",
      "59604  ENSG00000221983.7 9.94992918918919\n",
      "7134  ENSG00000142937.11 9.93833351351351\n",
      "38193 ENSG00000105290.11 9.93748810810811\n",
      "39006 ENSG00000172270.18 9.91861135135135\n",
      "                    gene      mean_Zscore\n",
      "55525 ENSG00000142089.15 9.99171401515152\n",
      "23072 ENSG00000104904.12 9.98586515151515\n",
      "60453 ENSG00000102265.11 9.98082954545455\n",
      "44413 ENSG00000187514.14 9.97383977272727\n",
      "15600  ENSG00000233927.4 9.93879128787879\n",
      "27705 ENSG00000173812.10 9.91041022727273\n",
      "                    gene      mean_Zscore\n",
      "9639  ENSG00000096384.19 9.99832942430704\n",
      "29063  ENSG00000248527.1  9.9758684434968\n",
      "14255 ENSG00000150991.14 9.96692132196162\n",
      "39006 ENSG00000172270.18 9.89217377398721\n",
      "15600  ENSG00000233927.4 9.86850341151386\n",
      "5847  ENSG00000149806.10 9.86291876332623\n",
      "                    gene      mean_Zscore\n",
      "24049 ENSG00000130726.11 9.97497532467533\n",
      "22949  ENSG00000219928.2 9.96885584415584\n",
      "42071  ENSG00000108518.7 9.95638506493506\n",
      "728   ENSG00000149925.16 9.95572532467532\n",
      "9639  ENSG00000096384.19 9.94919350649351\n",
      "39006 ENSG00000172270.18  9.9199461038961\n"
     ]
    }
   ],
   "source": [
    "cancer_top_1000 <- list()\n",
    "for(i in 1:length(tcga_projects)) {\n",
    "    top_1000 <- get.most.expressed(tcga_projects[i])\n",
    "    cancer_top_1000[[tcga_projects[i]]] <- top_1000\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1261cc06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "          Length Class  Mode     \n",
       "TCGA-BRCA 1000   -none- character\n",
       "TCGA-MESO  999   -none- character\n",
       "TCGA-CESC 1000   -none- character\n",
       "TCGA-BLCA 1000   -none- character\n",
       "TCGA-CHOL 1000   -none- character\n",
       "TCGA-ACC  1000   -none- character\n",
       "TCGA-DLBC 1000   -none- character\n",
       "TCGA-HNSC 1000   -none- character\n",
       "TCGA-COAD 1000   -none- character\n",
       "TCGA-ESCA 1000   -none- character\n",
       "TCGA-LAML  999   -none- character\n",
       "TCGA-KIRP 1000   -none- character\n",
       "TCGA-KIRC 1000   -none- character\n",
       "TCGA-GBM  1000   -none- character\n",
       "TCGA-KICH 1000   -none- character\n",
       "TCGA-READ 1000   -none- character\n",
       "TCGA-PAAD 1000   -none- character\n",
       "TCGA-LUAD 1000   -none- character\n",
       "TCGA-OV    999   -none- character\n",
       "TCGA-LIHC  999   -none- character\n",
       "TCGA-LUSC 1000   -none- character\n",
       "TCGA-LGG  1000   -none- character\n",
       "TCGA-UCEC 1000   -none- character\n",
       "TCGA-PRAD  998   -none- character\n",
       "TCGA-STAD 1000   -none- character\n",
       "TCGA-THCA 1000   -none- character\n",
       "TCGA-THYM 1000   -none- character\n",
       "TCGA-UCS  1000   -none- character\n",
       "TCGA-UVM  1000   -none- character\n",
       "TCGA-PCPG 1000   -none- character\n",
       "TCGA-SARC 1000   -none- character\n",
       "TCGA-SKCM 1000   -none- character\n",
       "TCGA-TGCT 1000   -none- character"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(cancer_top_1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "13de0929",
   "metadata": {},
   "outputs": [],
   "source": [
    "wrs_list <- readRDS(\"wrs_list.rds\")\n",
    "recurrent_wrs_list <- readRDS(\"recurrent_wrs_list.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9801e324",
   "metadata": {},
   "source": [
    "# Cancer enrichment Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "740fb1c5",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in summary(reccurent_wrs_list[[\"upregulated\"]]): object 'reccurent_wrs_list' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in summary(reccurent_wrs_list[[\"upregulated\"]]): object 'reccurent_wrs_list' not found\nTraceback:\n",
      "1. summary(reccurent_wrs_list[[\"upregulated\"]])"
     ]
    }
   ],
   "source": [
    "summary(reccurent_wrs_list[[\"upregulated\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fa90e29",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_type_up <- get_enrichment_data(wrs_list[[\"upregulated\"]],cancer_top_1000)\n",
    "cancer_type_up_new <- change_matrix(cancer_type_up$matrix, cancer_type_up$odds_matrix,2.5)\n",
    "cancer_type_up_new$barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe10ed46",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_type_down <- get_enrichment_data(wrs_list[[\"downregulated\"]],cancer_top_1000)\n",
    "cancer_type_down_new <- change_matrix(cancer_type_down$matrix, cancer_type_down$odds_matrix,4)\n",
    "cancer_type_down_new$barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6a24bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_type_up_reccurent <- get_enrichment_data(recurrent_wrs_list[[\"upregulated\"]],cancer_top_1000)\n",
    "cancer_type_up_reccurent_new <- change_matrix(cancer_type_up_reccurent$matrix, cancer_type_up_reccurent$odds_matrix,2.5)\n",
    "cancer_type_up_reccurent_new$barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8150a6d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(recurrent_wrs_list[[\"upregulated\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3ed083",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_type_down_reccurent <- get_enrichment_data(recurrent_wrs_list[[\"downregulated\"]],cancer_top_1000)\n",
    "cancer_type_down_reccurent_new <- change_matrix(cancer_type_down_reccurent$matrix, cancer_type_down_reccurent$odds_matrix,5)\n",
    "cancer_type_down_reccurent_new$barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "749820ae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1e569b7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R/3.6",
   "language": "R",
   "name": "ir36"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
