{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0dc9f79c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: usethis\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error in get(genname, envir = envir) : object 'testthat_print' not found\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.0 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.2     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.2     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4     \n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mbetween()\u001b[39m   masks \u001b[34mdata.table\u001b[39m::between()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m    masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m     masks \u001b[34mdata.table\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m       masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m      masks \u001b[34mdata.table\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m masks \u001b[34mdata.table\u001b[39m::transpose()\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(data.table)\n",
    "library(DescTools)\n",
    "library(ggplot2)\n",
    "library(devtools)\n",
    "library(pheatmap)\n",
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0d8a32d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"download completed in\"\n",
      "   user  system elapsed \n",
      " 19.225   8.894  36.187 \n"
     ]
    }
   ],
   "source": [
    "### files to be downloaded\n",
    "\n",
    "# read in the .txt file for tissue type as well as the larger gene expression file\\\n",
    "ptm <- proc.time()\n",
    "sample_attributes <- fread(file = \"/data/timonaj/gene_variability/GTEx_v7_Annotations_SampleAttributesDS.txt\")\n",
    "gene_tpm <- fread(file = \"/data/timonaj/gene_variability/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct\")\n",
    "gene_tpm_copy <- as.data.frame(gene_tpm)\n",
    "print(\"download completed in\")\n",
    "print(proc.time() - ptm)\n",
    "\n",
    "### commonly used variables\n",
    "\n",
    "# length of the numerical tpm values since the first 2 columns are characters\n",
    "tpm_length <- 3:length(gene_tpm_copy)\n",
    "\n",
    "tissue_types <- unique(sample_attributes$SMTS)\n",
    "\n",
    "gtex_project_files <- list()\n",
    "for(i in 1:length(tissue_types)) {\n",
    "  project <- tissue_types[i]\n",
    "  sampids<-sample_attributes[sample_attributes$SMTS == project,]$SAMPID\n",
    "  gtex_project_files[[project]] <- cbind(gene_tpm_copy[,1:2], gene_tpm_copy[,colnames(gene_tpm_copy) %in% sampids])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2fa2b9b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                Length Class      Mode\n",
       "Blood            539   data.frame list\n",
       "Adipose Tissue   799   data.frame list\n",
       "Muscle           566   data.frame list\n",
       "Blood Vessel     915   data.frame list\n",
       "Heart            602   data.frame list\n",
       "Ovary            135   data.frame list\n",
       "Uterus           113   data.frame list\n",
       "Vagina           117   data.frame list\n",
       "Breast           292   data.frame list\n",
       "Skin            1205   data.frame list\n",
       "Salivary Gland    99   data.frame list\n",
       "Brain           1673   data.frame list\n",
       "Adrenal Gland    192   data.frame list\n",
       "Thyroid          448   data.frame list\n",
       "Lung             429   data.frame list\n",
       "Spleen           164   data.frame list\n",
       "Pancreas         250   data.frame list\n",
       "Esophagus       1023   data.frame list\n",
       "Stomach          264   data.frame list\n",
       "Colon            509   data.frame list\n",
       "Small Intestine  139   data.frame list\n",
       "Prostate         154   data.frame list\n",
       "Testis           261   data.frame list\n",
       "Nerve            416   data.frame list\n",
       "Pituitary        185   data.frame list\n",
       "Liver            177   data.frame list\n",
       "Kidney            47   data.frame list\n",
       "Fallopian Tube     9   data.frame list\n",
       "Bladder           13   data.frame list\n",
       "Cervix Uteri      13   data.frame list\n",
       "Bone Marrow        2   data.frame list"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(gtex_project_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e43c4350",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"amexicanum_regen upregulated\"\n",
      "[1] \"amexicanum_regen total_datasets : 46\"\n",
      "[1] \"amexicanum_regen mean reccurence : 8.323\"\n",
      "[1] \"amexicanum_regen min reccurence : 5\"\n",
      "[1] \"amexicanum_regen max reccurence : 26\"\n",
      "[1] \"################################################\"\n",
      "[1] \"amexicanum_regen downregulated\"\n",
      "[1] \"amexicanum_regen total_datasets : 46\"\n",
      "[1] \"amexicanum_regen mean reccurence : 9.158\"\n",
      "[1] \"amexicanum_regen min reccurence : 5\"\n",
      "[1] \"amexicanum_regen max reccurence : 27\"\n",
      "[1] \"################################################\"\n",
      "[1] \"celegans_stress upregulated\"\n",
      "[1] \"celegans_stress total_datasets : 53\"\n",
      "[1] \"celegans_stress mean reccurence : 44.082\"\n",
      "[1] \"celegans_stress min reccurence : 38\"\n",
      "[1] \"celegans_stress max reccurence : 53\"\n",
      "[1] \"################################################\"\n",
      "[1] \"celegans_stress downregulated\"\n",
      "[1] \"celegans_stress total_datasets : 53\"\n",
      "[1] \"celegans_stress mean reccurence : 43.088\"\n",
      "[1] \"celegans_stress min reccurence : 36\"\n",
      "[1] \"celegans_stress max reccurence : 53\"\n",
      "[1] \"################################################\"\n",
      "[1] \"clupisFamiliaris_wound upregulated\"\n",
      "[1] \"clupisFamiliaris_wound total_datasets : 3\"\n",
      "[1] \"clupisFamiliaris_wound mean reccurence : 3\"\n",
      "[1] \"clupisFamiliaris_wound min reccurence : 3\"\n",
      "[1] \"clupisFamiliaris_wound max reccurence : 3\"\n",
      "[1] \"################################################\"\n",
      "[1] \"clupisFamiliaris_wound downregulated\"\n",
      "[1] \"clupisFamiliaris_wound total_datasets : 3\"\n",
      "[1] \"clupisFamiliaris_wound mean reccurence : 3\"\n",
      "[1] \"clupisFamiliaris_wound min reccurence : 3\"\n",
      "[1] \"clupisFamiliaris_wound max reccurence : 3\"\n",
      "[1] \"################################################\"\n",
      "[1] \"dmelanogaster_stress upregulated\"\n",
      "[1] \"dmelanogaster_stress total_datasets : 10\"\n",
      "[1] \"dmelanogaster_stress mean reccurence : 9.878\"\n",
      "[1] \"dmelanogaster_stress min reccurence : 9\"\n",
      "[1] \"dmelanogaster_stress max reccurence : 10\"\n",
      "[1] \"################################################\"\n",
      "[1] \"dmelanogaster_stress downregulated\"\n",
      "[1] \"dmelanogaster_stress total_datasets : 10\"\n",
      "[1] \"dmelanogaster_stress mean reccurence : 9.841\"\n",
      "[1] \"dmelanogaster_stress min reccurence : 9\"\n",
      "[1] \"dmelanogaster_stress max reccurence : 10\"\n",
      "[1] \"################################################\"\n",
      "[1] \"dmelanogaster_wound upregulated\"\n",
      "[1] \"dmelanogaster_wound total_datasets : 1\"\n",
      "[1] \"################################################\"\n",
      "[1] \"dmelanogaster_wound downregulated\"\n",
      "[1] \"dmelanogaster_wound total_datasets : 1\"\n",
      "[1] \"################################################\"\n",
      "[1] \"drerio_regen upregulated\"\n",
      "[1] \"drerio_regen total_datasets : 10\"\n",
      "[1] \"drerio_regen mean reccurence : 6.487\"\n",
      "[1] \"drerio_regen min reccurence : 5\"\n",
      "[1] \"drerio_regen max reccurence : 10\"\n",
      "[1] \"################################################\"\n",
      "[1] \"drerio_regen downregulated\"\n",
      "[1] \"drerio_regen total_datasets : 10\"\n",
      "[1] \"drerio_regen mean reccurence : 6.257\"\n",
      "[1] \"drerio_regen min reccurence : 5\"\n",
      "[1] \"drerio_regen max reccurence : 10\"\n",
      "[1] \"################################################\"\n",
      "[1] \"hsapiens_regen upregulated\"\n",
      "[1] \"hsapiens_regen total_datasets : 1\"\n",
      "[1] \"################################################\"\n",
      "[1] \"hsapiens_regen downregulated\"\n",
      "[1] \"hsapiens_regen total_datasets : 1\"\n",
      "[1] \"################################################\"\n",
      "[1] \"hsapiens_stress upregulated\"\n",
      "[1] \"hsapiens_stress total_datasets : 7\"\n",
      "[1] \"hsapiens_stress mean reccurence : 4.692\"\n",
      "[1] \"hsapiens_stress min reccurence : 3\"\n",
      "[1] \"hsapiens_stress max reccurence : 7\"\n",
      "[1] \"################################################\"\n",
      "[1] \"hsapiens_stress downregulated\"\n",
      "[1] \"hsapiens_stress total_datasets : 7\"\n",
      "[1] \"hsapiens_stress mean reccurence : 5.881\"\n",
      "[1] \"hsapiens_stress min reccurence : 5\"\n",
      "[1] \"hsapiens_stress max reccurence : 7\"\n",
      "[1] \"################################################\"\n",
      "[1] \"hsapiens_wound upregulated\"\n",
      "[1] \"hsapiens_wound total_datasets : 6\"\n",
      "[1] \"hsapiens_wound mean reccurence : 3.07537688442211\"\n",
      "[1] \"hsapiens_wound min reccurence : 3\"\n",
      "[1] \"hsapiens_wound max reccurence : 5\"\n",
      "[1] \"################################################\"\n",
      "[1] \"hsapiens_wound downregulated\"\n",
      "[1] \"hsapiens_wound total_datasets : 6\"\n",
      "[1] \"hsapiens_wound mean reccurence : 3.0084388185654\"\n",
      "[1] \"hsapiens_wound min reccurence : 3\"\n",
      "[1] \"hsapiens_wound max reccurence : 4\"\n",
      "[1] \"################################################\"\n",
      "[1] \"mmusculus_stress upregulated\"\n",
      "[1] \"mmusculus_stress total_datasets : 10\"\n",
      "[1] \"mmusculus_stress mean reccurence : 3.66943005181347\"\n",
      "[1] \"mmusculus_stress min reccurence : 3\"\n",
      "[1] \"mmusculus_stress max reccurence : 7\"\n",
      "[1] \"################################################\"\n",
      "[1] \"mmusculus_stress downregulated\"\n",
      "[1] \"mmusculus_stress total_datasets : 10\"\n",
      "[1] \"mmusculus_stress mean reccurence : 3.56914285714286\"\n",
      "[1] \"mmusculus_stress min reccurence : 3\"\n",
      "[1] \"mmusculus_stress max reccurence : 8\"\n",
      "[1] \"################################################\"\n",
      "[1] \"rnorvegicus_wound upregulated\"\n",
      "[1] \"rnorvegicus_wound total_datasets : 12\"\n",
      "[1] \"rnorvegicus_wound mean reccurence : 8.836\"\n",
      "[1] \"rnorvegicus_wound min reccurence : 7\"\n",
      "[1] \"rnorvegicus_wound max reccurence : 12\"\n",
      "[1] \"################################################\"\n",
      "[1] \"rnorvegicus_wound downregulated\"\n",
      "[1] \"rnorvegicus_wound total_datasets : 12\"\n",
      "[1] \"rnorvegicus_wound mean reccurence : 8.955\"\n",
      "[1] \"rnorvegicus_wound min reccurence : 7\"\n",
      "[1] \"rnorvegicus_wound max reccurence : 12\"\n",
      "[1] \"################################################\"\n",
      "[1] \"sscrofa_wound upregulated\"\n",
      "[1] \"sscrofa_wound total_datasets : 97\"\n",
      "[1] \"sscrofa_wound mean reccurence : 49.506\"\n",
      "[1] \"sscrofa_wound min reccurence : 42\"\n",
      "[1] \"sscrofa_wound max reccurence : 80\"\n",
      "[1] \"################################################\"\n",
      "[1] \"sscrofa_wound downregulated\"\n",
      "[1] \"sscrofa_wound total_datasets : 97\"\n",
      "[1] \"sscrofa_wound mean reccurence : 50.072\"\n",
      "[1] \"sscrofa_wound min reccurence : 42\"\n",
      "[1] \"sscrofa_wound max reccurence : 83\"\n",
      "[1] \"################################################\"\n"
     ]
    }
   ],
   "source": [
    "get_reccurent_genes <- function(duplicated_list,species_exptype,total_exps,title) {\n",
    "    print(paste(species_exptype,title, sep =\" \"))\n",
    "    \n",
    "    if(total_exps == 1) {\n",
    "        print(paste(species_exptype, \"total_datasets :\", total_exps, sep =\" \"))\n",
    "        print(\"################################################\")\n",
    "        return(duplicated_list)\n",
    "    }\n",
    "    \n",
    "    duplicated_list <- sort(table(duplicated_list), decreasing = TRUE)\n",
    "    \n",
    "    if(length(duplicated_list) > 1000) {\n",
    "        duplicated_list <- duplicated_list[1:1000]\n",
    "    }\n",
    "    \n",
    "    duplicated_list <- duplicated_list[duplicated_list > 2]\n",
    "    \n",
    "    print(paste(species_exptype, \"total_datasets :\", total_exps, sep =\" \"))\n",
    "    print(paste(species_exptype, \"mean reccurence :\", mean(duplicated_list), sep =\" \"))\n",
    "    print(paste(species_exptype, \"min reccurence :\", min(duplicated_list), sep =\" \"))\n",
    "    print(paste(species_exptype, \"max reccurence :\", max(duplicated_list), sep =\" \"))\n",
    "    print(\"################################################\")\n",
    "    return(names(duplicated_list))\n",
    "}\n",
    "\n",
    "wrs_list <- list(\"upregulated\" = list(),\n",
    "                 \"downregulated\" = list())\n",
    "recurrent_wrs_list <- list(\"upregulated\" = list(),\n",
    "                           \"downregulated\" = list())\n",
    "\n",
    "\n",
    "wrs_files <-list.files('./geo_degs/')\n",
    "wrs_foi <- wrs_files[grep(\"^[a-z].*\", wrs_files)]\n",
    "species_exptype <- unique(sub('_[A-Z].*$', '',wrs_foi))\n",
    "\n",
    "wrs_foi_up <- wrs_files[grep(\"^[a-z].*upregulated.*\", wrs_files)]\n",
    "wrs_foi_down <- wrs_files[grep(\"^[a-z].*downregulated.*\", wrs_files)]\n",
    "\n",
    "for(i in 1:length(species_exptype)) {\n",
    "    \n",
    "    current_species_up <- wrs_foi_up[grep(species_exptype[i], wrs_foi_up)]\n",
    "    current_species_down <- wrs_foi_down[grep(species_exptype[i], wrs_foi_down)]\n",
    "    total_spec_exp_up <- character(0)\n",
    "    total_spec_exp_down <- character(0)\n",
    "    \n",
    "    if(length(current_species_up) == length(current_species_down)) {\n",
    "        for(j in 1:length(current_species_up)) {\n",
    "            current_file_path_up <- paste('./geo_degs/', current_species_up[j], sep=\"\")\n",
    "            current_file_path_down <- paste('./geo_degs/', current_species_down[j], sep=\"\")\n",
    "            \n",
    "            total_genes_up <- fread(current_file_path_up, header=FALSE)$V1\n",
    "            total_genes_down <- fread(current_file_path_down, header=FALSE)$V1\n",
    "            \n",
    "            total_spec_exp_up <- append(total_spec_exp_up,\n",
    "                                        unique(total_genes_up),\n",
    "                                        length(total_spec_exp_up))\n",
    "            total_spec_exp_down <- append(total_spec_exp_down,\n",
    "                                          unique(total_genes_down),\n",
    "                                          length(total_spec_exp_down))\n",
    "        }\n",
    "    }\n",
    "    # union of all genes\n",
    "    wrs_list[[\"upregulated\"]][[species_exptype[i]]] <- unique(total_spec_exp_up)\n",
    "    wrs_list[[\"downregulated\"]][[species_exptype[i]]] <- unique(total_spec_exp_down)\n",
    "    \n",
    "    # top recurrent genes\n",
    "    recurrent_wrs_list[[\"upregulated\"]][[species_exptype[i]]] <- get_reccurent_genes(total_spec_exp_up,\n",
    "                                                                                     species_exptype[i],\n",
    "                                                                                     length(current_species_up),\n",
    "                                                                                     \"upregulated\")\n",
    "    recurrent_wrs_list[[\"downregulated\"]][[species_exptype[i]]] <- get_reccurent_genes(total_spec_exp_down,\n",
    "                                                                                       species_exptype[i],\n",
    "                                                                                       length(current_species_down),\n",
    "                                                                                     \"downregulated\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8316a64d",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in `[.data.frame`(cur_files, , 3:length(cur_files)): undefined columns selected\n",
     "output_type": "error",
     "traceback": [
      "Error in `[.data.frame`(cur_files, , 3:length(cur_files)): undefined columns selected\nTraceback:\n",
      "1. apply(cur_files[, 3:length(cur_files)], 1, mean)",
      "2. cur_files[, 3:length(cur_files)]",
      "3. `[.data.frame`(cur_files, , 3:length(cur_files))",
      "4. stop(\"undefined columns selected\")"
     ]
    }
   ],
   "source": [
    "gene_tpms <- data.frame(\"geneNames\" = gtex_project_files[[1]]$Description)\n",
    "for(i in 1:length(gtex_project_files)) {\n",
    "    cur_files <- gtex_project_files[[i]]\n",
    "    mean_tpm <- apply(cur_files[,3:length(cur_files)],1,mean)\n",
    "    gene_tpms <- cbind(gene_tpms, mean_tpm)\n",
    "    colnames(gene_tpms)[i+1] <- names(gtex_project_files)[i]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7fc425e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "ineq <- apply(gene_tpms[,2:length(gene_tpms)],\n",
    "      1,\n",
    "      function(x) {Gini(x,  n = rep(1, length(x)))})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "90d26bcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_speceficity <- cbind(\"genes\"= gene_tpms$geneNames,\n",
    "                            \"giniSpecificity\" = ineq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "848320c7",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error: $ operator is invalid for atomic vectors\n",
     "output_type": "error",
     "traceback": [
      "Error: $ operator is invalid for atomic vectors\nTraceback:\n",
      "1. tissue_speceficity$genes %in% current_list"
     ]
    }
   ],
   "source": [
    "total_scores <- numeric()\n",
    "spec_exptype <- character()\n",
    "for(i in 1:length(recurrent_wrs_list[[\"upregulated\"]])) {\n",
    "    current_list <- recurrent_wrs_list[[\"upregulated\"]][[i]]\n",
    "    scores <-tissue_speceficity[tissue_speceficity$genes %in% current_list,]$giniSpecificity\n",
    "    type <- names(recurrent_wrs_list[[\"upregulated\"]])[i]\n",
    "    spec_exptype <- append(spec_exptype, rep(type,length(scores)), length(spec_exptype))\n",
    "    total_scores <- append(total_scores, scores, length(total_scores))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b8ed67f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 31</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>geneNames</th><th scope=col>Blood</th><th scope=col>Adipose Tissue</th><th scope=col>Muscle</th><th scope=col>Blood Vessel</th><th scope=col>Heart</th><th scope=col>Ovary</th><th scope=col>Uterus</th><th scope=col>Vagina</th><th scope=col>Breast</th><th scope=col>⋯</th><th scope=col>Small Intestine</th><th scope=col>Prostate</th><th scope=col>Testis</th><th scope=col>Nerve</th><th scope=col>Pituitary</th><th scope=col>Liver</th><th scope=col>Kidney</th><th scope=col>Fallopian Tube</th><th scope=col>Bladder</th><th scope=col>Cervix Uteri</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>DDX11L1   </td><td>0.16160946</td><td> 0.07128650</td><td>0.08507681</td><td> 0.05952639</td><td>0.09960527</td><td> 0.04778271</td><td> 0.05215514</td><td> 0.04717713</td><td> 0.07792321</td><td>⋯</td><td> 0.07471299</td><td> 0.07388638</td><td> 1.79809305</td><td> 0.07047700</td><td> 0.05508787</td><td>0.08680291</td><td> 0.07558178</td><td> 0.04401714</td><td> 0.05492455</td><td> 0.05390182</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>WASH7P    </td><td>9.71775978</td><td>11.59039900</td><td>6.95123936</td><td>12.76671084</td><td>5.15145833</td><td>21.77142105</td><td>23.56507207</td><td>17.16219130</td><td>13.86089655</td><td>⋯</td><td>13.69509489</td><td>23.54824342</td><td>17.92720849</td><td>20.48243237</td><td>16.85879235</td><td>5.94239429</td><td>11.94446667</td><td>18.29714286</td><td>15.92909091</td><td>17.45054545</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>MIR1302-11</td><td>0.04362047</td><td> 0.08956950</td><td>0.11462793</td><td> 0.06773352</td><td>0.13767725</td><td> 0.04629594</td><td> 0.04648856</td><td> 0.04995313</td><td> 0.08464124</td><td>⋯</td><td> 0.08320533</td><td> 0.07748046</td><td> 0.08397089</td><td> 0.08764242</td><td> 0.05387372</td><td>0.10644160</td><td> 0.10659178</td><td> 0.05733571</td><td> 0.05916636</td><td> 0.04672545</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>FAM138A   </td><td>0.03001177</td><td> 0.05236898</td><td>0.06270426</td><td> 0.04245461</td><td>0.07623223</td><td> 0.02860692</td><td> 0.03460964</td><td> 0.02749096</td><td> 0.05363517</td><td>⋯</td><td> 0.04851949</td><td> 0.04858039</td><td> 0.04194907</td><td> 0.04572097</td><td> 0.03612760</td><td>0.05712463</td><td> 0.05535089</td><td> 0.04442286</td><td> 0.03293364</td><td> 0.04083909</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>OR4G4P    </td><td>0.01800994</td><td> 0.03783211</td><td>0.04583262</td><td> 0.02991457</td><td>0.05633497</td><td> 0.01776165</td><td> 0.02352423</td><td> 0.02718826</td><td> 0.04259466</td><td>⋯</td><td> 0.03677350</td><td> 0.03632553</td><td> 0.03234367</td><td> 0.03593203</td><td> 0.02602055</td><td>0.04110366</td><td> 0.04303200</td><td> 0.00000000</td><td> 0.03350636</td><td> 0.03863909</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>OR4G11P   </td><td>0.02715438</td><td> 0.05627418</td><td>0.07348340</td><td> 0.04631588</td><td>0.08750470</td><td> 0.02784083</td><td> 0.03651315</td><td> 0.03272930</td><td> 0.05658034</td><td>⋯</td><td> 0.05316737</td><td> 0.05122507</td><td> 0.06487189</td><td> 0.04791353</td><td> 0.03921514</td><td>0.06607046</td><td> 0.06330889</td><td> 0.04937429</td><td> 0.02537091</td><td> 0.03379273</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 31\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & geneNames & Blood & Adipose Tissue & Muscle & Blood Vessel & Heart & Ovary & Uterus & Vagina & Breast & ⋯ & Small Intestine & Prostate & Testis & Nerve & Pituitary & Liver & Kidney & Fallopian Tube & Bladder & Cervix Uteri\\\\\n",
       "  & <fct> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & DDX11L1    & 0.16160946 &  0.07128650 & 0.08507681 &  0.05952639 & 0.09960527 &  0.04778271 &  0.05215514 &  0.04717713 &  0.07792321 & ⋯ &  0.07471299 &  0.07388638 &  1.79809305 &  0.07047700 &  0.05508787 & 0.08680291 &  0.07558178 &  0.04401714 &  0.05492455 &  0.05390182\\\\\n",
       "\t2 & WASH7P     & 9.71775978 & 11.59039900 & 6.95123936 & 12.76671084 & 5.15145833 & 21.77142105 & 23.56507207 & 17.16219130 & 13.86089655 & ⋯ & 13.69509489 & 23.54824342 & 17.92720849 & 20.48243237 & 16.85879235 & 5.94239429 & 11.94446667 & 18.29714286 & 15.92909091 & 17.45054545\\\\\n",
       "\t3 & MIR1302-11 & 0.04362047 &  0.08956950 & 0.11462793 &  0.06773352 & 0.13767725 &  0.04629594 &  0.04648856 &  0.04995313 &  0.08464124 & ⋯ &  0.08320533 &  0.07748046 &  0.08397089 &  0.08764242 &  0.05387372 & 0.10644160 &  0.10659178 &  0.05733571 &  0.05916636 &  0.04672545\\\\\n",
       "\t4 & FAM138A    & 0.03001177 &  0.05236898 & 0.06270426 &  0.04245461 & 0.07623223 &  0.02860692 &  0.03460964 &  0.02749096 &  0.05363517 & ⋯ &  0.04851949 &  0.04858039 &  0.04194907 &  0.04572097 &  0.03612760 & 0.05712463 &  0.05535089 &  0.04442286 &  0.03293364 &  0.04083909\\\\\n",
       "\t5 & OR4G4P     & 0.01800994 &  0.03783211 & 0.04583262 &  0.02991457 & 0.05633497 &  0.01776165 &  0.02352423 &  0.02718826 &  0.04259466 & ⋯ &  0.03677350 &  0.03632553 &  0.03234367 &  0.03593203 &  0.02602055 & 0.04110366 &  0.04303200 &  0.00000000 &  0.03350636 &  0.03863909\\\\\n",
       "\t6 & OR4G11P    & 0.02715438 &  0.05627418 & 0.07348340 &  0.04631588 & 0.08750470 &  0.02784083 &  0.03651315 &  0.03272930 &  0.05658034 & ⋯ &  0.05316737 &  0.05122507 &  0.06487189 &  0.04791353 &  0.03921514 & 0.06607046 &  0.06330889 &  0.04937429 &  0.02537091 &  0.03379273\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 31\n",
       "\n",
       "| <!--/--> | geneNames &lt;fct&gt; | Blood &lt;dbl&gt; | Adipose Tissue &lt;dbl&gt; | Muscle &lt;dbl&gt; | Blood Vessel &lt;dbl&gt; | Heart &lt;dbl&gt; | Ovary &lt;dbl&gt; | Uterus &lt;dbl&gt; | Vagina &lt;dbl&gt; | Breast &lt;dbl&gt; | ⋯ ⋯ | Small Intestine &lt;dbl&gt; | Prostate &lt;dbl&gt; | Testis &lt;dbl&gt; | Nerve &lt;dbl&gt; | Pituitary &lt;dbl&gt; | Liver &lt;dbl&gt; | Kidney &lt;dbl&gt; | Fallopian Tube &lt;dbl&gt; | Bladder &lt;dbl&gt; | Cervix Uteri &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | DDX11L1    | 0.16160946 |  0.07128650 | 0.08507681 |  0.05952639 | 0.09960527 |  0.04778271 |  0.05215514 |  0.04717713 |  0.07792321 | ⋯ |  0.07471299 |  0.07388638 |  1.79809305 |  0.07047700 |  0.05508787 | 0.08680291 |  0.07558178 |  0.04401714 |  0.05492455 |  0.05390182 |\n",
       "| 2 | WASH7P     | 9.71775978 | 11.59039900 | 6.95123936 | 12.76671084 | 5.15145833 | 21.77142105 | 23.56507207 | 17.16219130 | 13.86089655 | ⋯ | 13.69509489 | 23.54824342 | 17.92720849 | 20.48243237 | 16.85879235 | 5.94239429 | 11.94446667 | 18.29714286 | 15.92909091 | 17.45054545 |\n",
       "| 3 | MIR1302-11 | 0.04362047 |  0.08956950 | 0.11462793 |  0.06773352 | 0.13767725 |  0.04629594 |  0.04648856 |  0.04995313 |  0.08464124 | ⋯ |  0.08320533 |  0.07748046 |  0.08397089 |  0.08764242 |  0.05387372 | 0.10644160 |  0.10659178 |  0.05733571 |  0.05916636 |  0.04672545 |\n",
       "| 4 | FAM138A    | 0.03001177 |  0.05236898 | 0.06270426 |  0.04245461 | 0.07623223 |  0.02860692 |  0.03460964 |  0.02749096 |  0.05363517 | ⋯ |  0.04851949 |  0.04858039 |  0.04194907 |  0.04572097 |  0.03612760 | 0.05712463 |  0.05535089 |  0.04442286 |  0.03293364 |  0.04083909 |\n",
       "| 5 | OR4G4P     | 0.01800994 |  0.03783211 | 0.04583262 |  0.02991457 | 0.05633497 |  0.01776165 |  0.02352423 |  0.02718826 |  0.04259466 | ⋯ |  0.03677350 |  0.03632553 |  0.03234367 |  0.03593203 |  0.02602055 | 0.04110366 |  0.04303200 |  0.00000000 |  0.03350636 |  0.03863909 |\n",
       "| 6 | OR4G11P    | 0.02715438 |  0.05627418 | 0.07348340 |  0.04631588 | 0.08750470 |  0.02784083 |  0.03651315 |  0.03272930 |  0.05658034 | ⋯ |  0.05316737 |  0.05122507 |  0.06487189 |  0.04791353 |  0.03921514 | 0.06607046 |  0.06330889 |  0.04937429 |  0.02537091 |  0.03379273 |\n",
       "\n"
      ],
      "text/plain": [
       "  geneNames  Blood      Adipose Tissue Muscle     Blood Vessel Heart     \n",
       "1 DDX11L1    0.16160946  0.07128650    0.08507681  0.05952639  0.09960527\n",
       "2 WASH7P     9.71775978 11.59039900    6.95123936 12.76671084  5.15145833\n",
       "3 MIR1302-11 0.04362047  0.08956950    0.11462793  0.06773352  0.13767725\n",
       "4 FAM138A    0.03001177  0.05236898    0.06270426  0.04245461  0.07623223\n",
       "5 OR4G4P     0.01800994  0.03783211    0.04583262  0.02991457  0.05633497\n",
       "6 OR4G11P    0.02715438  0.05627418    0.07348340  0.04631588  0.08750470\n",
       "  Ovary       Uterus      Vagina      Breast      ⋯ Small Intestine Prostate   \n",
       "1  0.04778271  0.05215514  0.04717713  0.07792321 ⋯  0.07471299      0.07388638\n",
       "2 21.77142105 23.56507207 17.16219130 13.86089655 ⋯ 13.69509489     23.54824342\n",
       "3  0.04629594  0.04648856  0.04995313  0.08464124 ⋯  0.08320533      0.07748046\n",
       "4  0.02860692  0.03460964  0.02749096  0.05363517 ⋯  0.04851949      0.04858039\n",
       "5  0.01776165  0.02352423  0.02718826  0.04259466 ⋯  0.03677350      0.03632553\n",
       "6  0.02784083  0.03651315  0.03272930  0.05658034 ⋯  0.05316737      0.05122507\n",
       "  Testis      Nerve       Pituitary   Liver      Kidney      Fallopian Tube\n",
       "1  1.79809305  0.07047700  0.05508787 0.08680291  0.07558178  0.04401714   \n",
       "2 17.92720849 20.48243237 16.85879235 5.94239429 11.94446667 18.29714286   \n",
       "3  0.08397089  0.08764242  0.05387372 0.10644160  0.10659178  0.05733571   \n",
       "4  0.04194907  0.04572097  0.03612760 0.05712463  0.05535089  0.04442286   \n",
       "5  0.03234367  0.03593203  0.02602055 0.04110366  0.04303200  0.00000000   \n",
       "6  0.06487189  0.04791353  0.03921514 0.06607046  0.06330889  0.04937429   \n",
       "  Bladder     Cervix Uteri\n",
       "1  0.05492455  0.05390182 \n",
       "2 15.92909091 17.45054545 \n",
       "3  0.05916636  0.04672545 \n",
       "4  0.03293364  0.04083909 \n",
       "5  0.03350636  0.03863909 \n",
       "6  0.02537091  0.03379273 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gene_tpms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "835a4f76",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R/3.6",
   "language": "R",
   "name": "ir36"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
